ENH: Implemented get_symmetrised_molecule + helper functions; Improved performance of generate_full_symmops

mcocdawc opened the pull request:
# Summary
Closes materialsproject/pymatgen#746

New public functions:
* ``PointGroupAnalyzer.symmetrize_molecule``
* ``PointGroupAnalyzer.get_equivalent_atoms``

Changes in public functions:
* ``analyzer.cluster_sites`` now supports an optional boolean keyword argument ``give_only_index``, which defaults to ``False`` (previous behaviour)
* ``analyze.generate_full_symmops`` was optimised numerically and does not use recursion anymore. In the case of icosahedral point groups (tested on fullerene) the new implementation is 72 times faster.

New private functions (helper functions for ``PointGroupAnalyzer.get_equivalent_atoms``)
* ``PointGroupAnalyzer._get_eq_sets``
* ``PointGroupAnalyzer._combine_eq_sets``

shyuep commented:
Thanks! Do you mind adding a simple unittest? E.g., give it a slightly distorted methane molecule and symmetrize it.

mcocdawc commented:
The tests are not yet finished. I will finish them tomorrow

mcocdawc commented:
Tests are finished and I added some more functionality:
Added public function:
``iterative_symmetrize`` applies ``PointGroupAnalyzer.symmetrize_molecule`` in a self consistent manner.
Since this requires the creation of new ``PointGroupAnalyzer`` instances, I didn't make it a method, but a function.
From my side, I would say it is finished.

shyuep commented:
Thanks a lot!

mcocdawc commented:
Thanks a lot for the project. ;)

mcocdawc commented:
@shyuep May I ask, how often the PyPi package gets updated, because I would like to use ``iterative_symmetrize`` as dependency in a package of mine.

shyuep commented:
Ah, ok. I can release a version today. There is usually no fixed timeframe to releases. I use my judgment on when a collection of bug fixes or features have reached a critical mass for a release.

mcocdawc commented:
Thank you very much.
From my side there is definetely no hurry to update it in the timeframe of days.
I just wanted to make sure, that you don't have release cycles of half a year.

shyuep commented:
Ok, I just released a new version.

Oh, don't worry. Releases definitely happen more than once a month typically. Scientific research cannot wait half a year for releases! ;-)

mcocdawc commented:
Yeah makes sense. ;)
